hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/UserMetricsContext.java

```diff
@@ -0,0 +1,208 @@
+import com.hazelcast.internal.metrics.ProbeLevel;
+import com.hazelcast.internal.metrics.ProbeUnit;
+import com.hazelcast.jet.core.metrics.Counter;
+import com.hazelcast.jet.core.metrics.Gauge;
+import com.hazelcast.jet.core.metrics.Unit;
+import com.hazelcast.jet.impl.execution.ProcessorTasklet;
+
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+public class UserMetricsContext {
+
+    private final ProbeBuilder probeBuilder;
+    private final ProcessorTasklet source;
+    private final ProbeLevel level;
+
+    private String onlyName;
+    private Metric onlyMetric;
+    private Map<String, Metric> metrics;
+
+    public UserMetricsContext(ProbeBuilder probeBuilder, ProcessorTasklet source, ProbeLevel level) {
+        this.probeBuilder = probeBuilder;
+        this.source = source;
+        this.level = level;
+    }
+
+    Counter getCounter(String name) {
+        return getMetric(name, Unit.COUNT);
+    }
+
+    Gauge getGauge(String name, Unit unit) {
+        return getMetric(name, unit);
+    }
+
+    private Metric getMetric(String name, Unit unit) {
+        Metric res = null;
+        if (name.equals(onlyName)) {
+            res = onlyMetric;
+        } else if (metrics != null) {
+            res = metrics.get(name);
+        }
+        // register metric on first use
+        if (res == null) {
+            res = initMetric(name, unit);
+            if (onlyName == null) {
+                // the first and so far the only metric
+                onlyName = name;
+                onlyMetric = res;
+            } else {
+                // 2 or more metrics
+                if (metrics == null) {
+                    metrics = new HashMap<>();
+                    metrics.put(onlyName, onlyMetric);
+                    onlyName = null;
+                    onlyMetric = null;
+                }
+                metrics.put(name, res);
+            }
+        }
+        return res;
+    }
+
+    private Metric initMetric(String name, Unit unit) {
+        if (probeBuilder == null) {
+            // used when metrics are disabled
+            return new NonRegisteredMetric(name);
+        } else {
+            RegisteredMetric metric = new RegisteredMetric(name);
+            LongProbeFunction<ProcessorTasklet> longProbeFunction = t -> metric.get();
+            probeBuilder.register(source, name, level, toProbeUnit(unit), longProbeFunction);
+            return metric;
+        }
+    }
+
+    private ProbeUnit toProbeUnit(Unit unit) {
+        switch (unit) {
+            case COUNT:
+                return ProbeUnit.COUNT;
+            case MS:
+                return ProbeUnit.MS;
+            case BYTES:
+                return ProbeUnit.BYTES;
+            default:
+                throw new RuntimeException("Unhandled metrics unit " + unit);
+        }
+    }
+
+    private interface Metric extends Counter, Gauge {
```

Comment:
this interface should be removed

Reply:
Why is it a problem? It's private and simplifies the code in this class a great deal.

Reply:
I thought the class impl is simple enough we don't need two different interfaces for it.